datafusion/physical-plan/src/joins/ie_join.rs

```diff
@@ -0,0 +1,1160 @@
+                Arc::new(permutation),
+            ],
+            vec![
+                // order by condition 2
+                (0, sort_options[1]),
+                (
+                    1,
+                    SortOptions {
+                        // same as above
+                        descending: !is_loose_inequality_operator(&operators[1]),
+                        nulls_first: false,
+                    },
+                ),
+            ],
+        );
+        l2.sort_by_columns()?;
+        let valid = (l2.arrays()[0].len() - l2.arrays()[0].null_count()) as usize;
+        let l2 = l2.slice(0..valid);
+
+        Ok((
+            l1_indexes,
+            Arc::clone(&l2.arrays()[2])
+                .as_primitive::<UInt64Type>()
+                .clone(),
+        ))
+    }
+
+    /// compute the join indices statify the inequality conditions
+    /// following the example in `compute_permutation`, the l1_indexes is \[1, -1, 2, -2\], the permutation is \[2, 1, 3, 0\]
+    /// range_map is empty at first
+    /// 1、 p\[0\] = 2, range_map is empty, l1_indexes\[2\] is greater than 0, it means 2nd element in l1 is from left table, insert(2) into range_map, range_map {(2, 3)}
+    /// 2、 p\[1\] = 1, no value less than p\[1\] in range_map, l1_indexes\[1\] is less than 0, it means 1st element in l1 is from right table, no need to insert(1) into range_map, range_map {(2, 3)}
+    /// 3、 p\[2\] = 3, found 2 less than p\[2\] in range_map, append all pairs (l1_indexes\[2\], l1_indexes\[3\]) to the indeices array, l1_indexes\[3\] is less than 0, it means 3rd element in l1 is from right table, no need to insert(3) into range_map, range_map {(1, 4)}
+    /// 4、 p\[3\] = 0, no value less than p\[3\] in range_map, insert(0) into range_map, range_map {(0, 1), (2, 3)}
+    /// The indices array is \[(2), (2)\]
+    fn build_join_indices(
+        l1_indexes: &Int64Array,
+        permutation: &UInt64Array,
+    ) -> Result<(UInt64Array, UInt64Array)> {
+        let mut left_builder = UInt64Builder::new();
+        let mut right_builder = UInt64Builder::new();
+        // left_order\[i\] = l means there are l elements from left table in l1\[0..=i\], also means element i is the l-th smallest element in left recordbatch.
+        let mut left_order = UInt64Array::builder(l1_indexes.len());
+        let mut l_pos = 0;
+        for ind in l1_indexes.values().iter() {
+            if *ind < 0 {
+                l_pos += 1;
+            }
+            left_order.append_value(l_pos);
+        }
+        let left_order = left_order.finish();
+        // use btree map to maintain all p\[i\], for i in 0..j, map\[s\]=t means range \[s, t\) is valid
+        // our target is to find all pair(i, j) that i<j and p\[i\] < p\[j\] and i from left table and j from right table here
+        // range_map use key as end index and value as start index to represent a interval [start, end)
+        let mut range_map = BTreeMap::<u64, u64>::new();
```

Comment:
In my test, use btreemap is a little faster than bitmap (8.5s -> 7.5s). Currently, the main cost is sorting though.